src/pool/srv_pool.c

```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST) {
+		val = DAOS_PROP_PO_RP_PDA_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_rp_pda, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade RP performance domain "
+				"affinity of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	d_iov_set(&value, &val32, sizeof(val32));
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_upgrade_status, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST || val32 != DAOS_UPGRADE_STATUS_IN_PROGRESS) {
+		val32 = DAOS_UPGRADE_STATUS_IN_PROGRESS;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_upgrade_status, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade 'upgrade status' "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_upgrade_global_version,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST || val32 != DS_POOL_GLOBAL_VERSION) {
+		val32 = DS_POOL_GLOBAL_VERSION;
+		rc = rdb_tx_update(tx, &svc->ps_root,
+				   &ds_pool_prop_upgrade_global_version, &value);
+		if (rc != 0) {
+			D_ERROR("failed to write upgrade global version prop, "DF_RC"\n",
+				DP_RC(rc));
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	if (need_commit) {
+		daos_prop_t *prop = NULL;
+
+		rc = rdb_tx_commit(tx);
+		if (rc)
+			D_GOTO(out_free, rc);
+		rc = pool_prop_read(tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
+		if (rc)
+			D_GOTO(out_free, rc);
+		rc = ds_pool_iv_prop_update(svc->ps_pool, prop);
+	}
+
+out_free:
+	D_FREE(hdl_uuids);
+	return rc;
+}
+
+static int ds_pool_mark_upgrade_completed(uuid_t pool_uuid,
+					  struct pool_svc *svc, int rc)
+{
+	struct rdb_tx			tx;
+	d_iov_t				value;
+	uint32_t			val;
+	int				rc1;
+	bool				need_put_leader = false;
+	daos_prop_t			*prop = NULL;
+
+	if (!svc) {
+		rc1 = pool_svc_lookup_leader(pool_uuid, &svc, NULL);
+		if (rc1 != 0)
+			return rc1;
+	}
+
+	rc1 = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc1 != 0)
+		D_GOTO(out_svc, rc1);
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+	if (rc == 0)
+		val = DAOS_UPGRADE_STATUS_COMPLETED;
+	else
+		val = DAOS_UPGRADE_STATUS_FAILED;
+
+	d_iov_set(&value, &val, sizeof(val));
+	rc1 = rdb_tx_update(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+			   &value);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+
+	/* only bump global version if upgrade succeed. */
+	if (rc == 0) {
+		val = DS_POOL_GLOBAL_VERSION;
+		rc1 = rdb_tx_update(&tx, &svc->ps_root,
+				    &ds_pool_prop_global_version, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade global version "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc1);
+			D_GOTO(out_tx, rc1);
+		}
+	}
+
+	rc1 = pool_prop_read(&tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
```

Comment:
Note that an RDB TX can't read its own uncommitted writes. See https://github.com/daos-stack/daos/blob/629ca9714a01d83f2dbd53a3a28de09afb81f597/src/include/daos_srv/rdb.h#L63